### PR TITLE
[Feature] Constant Interpolated Strings- Versioning Check & Test Suite

### PR DESCRIPTION
**Pull request by kevinsun-dev**
## [Feature] Constant Interpolated Strings

New in this PR:
- Versioning Check for Constant Interpolated Strings
- Updated Tests

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs [hunk 1]
```diff
@@ -155,5 +157,21 @@ internal static string UnescapeInterpolatedStringLiteral(string s)
 
             return value;
         }
+
+        private sealed class CheckConstantInterpolatedStringValidity : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+        {
+            internal DiagnosticBag diagnostics;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:163]: 'readonly'?

Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:163]: Added to next commit. Anything else?

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs [hunk 2]
```diff
@@ -155,5 +157,21 @@ internal static string UnescapeInterpolatedStringLiteral(string s)
...
+            public CheckConstantInterpolatedStringValidity(DiagnosticBag diagnostics)
+            {
+                this.diagnostics = diagnostics;
+            }
+
+            public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
+            {
+                Binder.CheckFeatureAvailability(node.Syntax, MessageID.IDS_FeatureConstantInterpolatedStrings, diagnostics);
+                return base.VisitInterpolatedString(node);
+            }
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:174]: We don't need to continue recursion. Just return `null` and it's good.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 3]
```diff
@@ -3541,6 +3541,103 @@ void M()
...
+                //         const string S5 = "Hybrid" + "Testing" + $"321";
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, @"$""321""").WithArguments("constant interpolated strings").WithLocation(10, 50),
+                // (11,27): error CS8652: The feature 'constant interpolated strings' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         const string F1 = $"{S1}";
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, @"$""{S1}""").WithArguments("constant interpolated strings").WithLocation(11, 27),
+                // (12,32): error CS8652: The feature 'constant interpolated strings' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //         const string F2 = F1 + $" the {S2}";
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, @"$"" the {S2}""").WithArguments("constant interpolated strings").WithLocation(12, 32));
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3639]: Please add some tests for:

* Parameter default values
* When the langversion is too low _and_ the string isn't a constant.

For each of the tests around langversion, our general strategy is to compile twice. Once with the correct minimum langversion to show it works, then once with the max unsupported langversion, to show that it doesn't.

**Comment by 333fred** [comment]
Done review pass (commit 7)

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs (hunk 2 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:174]: Thanks for the reminder! Fixed

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs (hunk 3 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs]: Added tests, let me know what you think

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs [hunk 4]
```diff
@@ -155,5 +157,21 @@ internal static string UnescapeInterpolatedStringLiteral(string s)
 
             return value;
         }
+
+        private sealed class CheckConstantInterpolatedStringValidity : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
+        {
+            readonly internal DiagnosticBag diagnostics;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:163]: Visibility modifiers should always come first.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 5]
```diff
@@ -3515,31 +3515,146 @@ public void ConstantInterpolatedStringsError()
...
+        const string S3 = $""{$""{""Spinning Top""}""}"";
+        const string S4 = $""Hybrid"" + ""Testing"" + ""123"";
+        const string S5 = ""Hybrid"" + ""Testing"" + $""321"";
+        const string F1 = $""{S1}"";
+        const string F2 = F1 + $"" the {S2}"";
+
+        string VS = ""Change"";
+        const string S6 = $""Failed to {VS}"";
+
+        // Remove unused variable warnings from diagnostics.
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3619]: Prefer to just use `pragma warning disable CSXXXX // unused locals` instead.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 6]
```diff
@@ -3515,31 +3515,146 @@ public void ConstantInterpolatedStringsError()
             string source = @"
 class C
 {
-    void M()
+    void M(string ParamDefault = ""Academy City"")
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3518]: >""Academy City"" [](start = 33, length = 16)

This is not what I meant by parameter default. I meant using an interpolated string in this location. Please break that out into a separate test.

**Comment by 333fred** [comment]
Done review pass (commit 8)

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs (hunk 4 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs]: Fixed, thanks!

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs (hunk 5 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs]: Added, thanks for the tip

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs (hunk 6 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3536]: Ah, that makes sense. Broke out a new test for interpolated string default, and also tested for use of constant in switch statement.

Comment (333fred) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3536]: We need to test langversion for these locations :). Sorry

**Comment by 333fred** [comment]
Done review pass (commit 9). Just a couple more tests.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 7]
```diff
@@ -3509,37 +3509,176 @@ void M()
...
+        const string S4 = $""Hybrid"" + ""Testing"" + ""123"";
+        const string S5 = ""Hybrid"" + ""Testing"" + $""321"";
+        const string F1 = $""{S1}"";
+        const string F2 = F1 + $"" the {S2}"";
+
+        string VS = ""Change"";
+        const string S6 = $""Failed to {VS}"";
+    }
+
+    void M2(string S1 = $""Testing"")
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3639]: Why is this test succeeding? I would expect to see failures in version 8 for both this location and the one below.

If you want to just mark them with a prototype comment and address in a follow-up, that's fine. But there needs to be an error :)

Comment (kevinsun-dev) [reply src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3639]: Huh, you're absolutely right. Those shouldn't be as they are. I suspect there's a separate path that checks for constants in those cases, or maybe it was somewhere higher up the chain. I'll take a look. If I can't find it by noon, I'll slap a PROTOTYPE on it.

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs [hunk 8]
```diff
@@ -155,5 +157,21 @@ internal static string UnescapeInterpolatedStringLiteral(string s)
 
             return value;
         }
+
+        internal class CheckConstantInterpolatedStringValidity : BoundTreeWalkerWithStackGuardWithoutRecursionOnTheLeftOfBinaryOperator
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:161]: Should still be `sealed`

File: src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs [hunk 9]
```diff
@@ -289,6 +289,11 @@ private ExpressionSyntax SkipParensAndNullSuppressions(ExpressionSyntax e)
                 {
                     CheckFeatureAvailability(patternExpression, MessageID.IDS_FeatureNullPointerConstantPattern, diagnostics, patternExpression.Location);
                 }
+                else if (inputType.IsStringType())
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:292]: >inputType.IsStringType() [](start = 25, length = 24)

I don't believe this is correct. Please add a test where the `o` is of type `object` in `switch (o) { case $"test": break; }`

Comment (333fred) [reply src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs:292]: It should be checking to see whether the pattern itself has a constant value of type string.

---
In reply to: [453030193](https://github.com/dotnet/roslyn/pull/45685#discussion_r453030193) [](ancestors = 453030193)

**Comment by 333fred** [comment]
Done review pass (commit 11)

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs (hunk 8 above)
Comment (kevinsun-dev) [reply src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:161]: I don't think so. It seems that sealed makes it inaccessible from some of the other files I'm calling it in.

File: src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs [hunk 10]
```diff
@@ -155,5 +156,30 @@ internal static string UnescapeInterpolatedStringLiteral(string s)
...
+            }
+
+            public override BoundNode VisitInterpolatedString(BoundInterpolatedString node)
+            {
+                Binder.CheckFeatureAvailability(node.Syntax, MessageID.IDS_FeatureConstantInterpolatedStrings, diagnostics);
+                return null;
+            }
+        }
+
+        internal static void CheckLangVersionForConstantInterpolatedStrings(BoundExpression expression, DiagnosticBag diagnostics)
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/ConstantValueUtils.cs:176]: >CheckLangVersionForConstantInterpolatedStrings [](start = 29, length = 46)

I would be less specific here. Just `CheckLangVersionForConstantValue`. That way, if we do anything else in this area in the future with non-strings, we have an obvious place to put it.

**Comment by 333fred** [comment]
Done review pass (commit 12). One more small rename and I think this should be good.

**Comment by 333fred** [comment]
Test failures appear legitimate.

File: Compilers.sln [hunk 11]
```diff
@@ -62,7 +62,7 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CSharpErrorFactsGenerator",
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CSharpSyntaxGenerator", "src\Tools\Source\CompilerGeneratorTools\Source\CSharpSyntaxGenerator\CSharpSyntaxGenerator.csproj", "{288089C5-8721-458E-BE3E-78990DAB5E2D}"
 EndProject
-Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "IOperationGenerator", "src\Tools\Source\CompilerGeneratorTools\Source\IOperationGenerator\CompilersIOperationGenerator.csproj", "{D0A79850-B32A-45E5-9FD5-D43CB345867A}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "CompilersIOperationGenerator", "src\Tools\Source\CompilerGeneratorTools\Source\IOperationGenerator\CompilersIOperationGenerator.csproj", "{D0A79850-B32A-45E5-9FD5-D43CB345867A}"
```
Comment (333fred) [inline Compilers.sln:65]: These changes seem unrelated.

Comment (kevinsun-dev) [reply Compilers.sln:65]: I don't quite know how that got in there, I thought I explicitly left it out, like I had for the past several commits. Does it make enough of a difference to remove it?

Comment (333fred) [reply Compilers.sln:65]: Yes, please remove it.

Comment (kevinsun-dev) [reply Compilers.sln]: Huh, git checkout can revert just a single file. That's a really convenient trick.

**Comment by 333fred** [comment]
@dotnet/roslyn-compiler for a second review.

File: src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs [hunk 12]
```diff
@@ -3509,37 +3509,225 @@ void M()
...
+        switch(S1){
+            case $""Level 5"":
+                break;
+            case $""Radio Noise"":
+                goto case $""Level 5"";
+        }
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.RegularPreview);
+            comp.VerifyDiagnostics();
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Semantic/Semantics/ConstantTests.cs:3540]: >VerifyDiagnostics [](start = 17, length = 17)

Can you also please verify diagnostics for a C# 8 compilation?
